pkg/ds/farm.go

```diff
@@ -0,0 +1,358 @@
+	alerter alerting.Alerter,
+) *Farm {
+	if alerter == nil {
+		alerter = alerting.NewNop()
+	}
+
+	return &Farm{
+		kpStore:    kpStore,
+		dsStore:    dsStore,
+		scheduler:  scheduler.NewApplicatorScheduler(applicator),
+		applicator: applicator,
+		children:   make(map[fields.ID]*childDS),
+		logger:     logger,
+		alerter:    alerter,
+	}
+}
+
+func (dsf *Farm) Start(quitCh <-chan struct{}) {
+	dsf.cleanupDaemonSetPods(quitCh)
+	dsf.mainLoop(quitCh)
+}
+
+const cleanupInterval = 60 * time.Second
+
+// This function removes all pods with a DSIDLabel where the daemon set id does
+// not exist in the store at every interval specified because it is possible
+// that the farm will unexpectedly crash or someone deletes or modifies a node
+func (dsf *Farm) cleanupDaemonSetPods(quitCh <-chan struct{}) {
+	go func() {
+		timer := time.NewTimer(time.Duration(0))
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			timer.Reset(cleanupInterval)
+
+			allDaemonSets, err := dsf.dsStore.List()
+			dsIDMap := make(map[string]ds_fields.DaemonSet)
+			for _, dsFields := range allDaemonSets {
+				dsIDMap[dsFields.ID.String()] = dsFields
+			}
+
+			allPods, err := dsf.applicator.GetMatches(klabels.Everything(), labels.POD)
+			if err != nil {
+				dsf.logger.NoFields().Fatal(err)
+				return
+			}
+
+			for _, podLabels := range allPods {
+				// Only check if it is a pod scheduled by a daemon set
+				if podLabels.Labels.Has(DSIDLabel) {
+					dsID := podLabels.Labels.Get(DSIDLabel)
+
+					// Check if the daemon set exists, if it doesn't unschedule the pod
+					if _, ok := dsIDMap[dsID]; !ok {
+						nodeName, podID, err := labels.NodeAndPodIDFromPodLabel(podLabels)
+						if err != nil {
+							dsf.logger.NoFields().Fatal(err)
+							return
+						}
+
+						// TODO: Since this mirrors the unschedule function in daemon_set.go,
+						// We should find a nice way to couple them together
+						dsf.logger.NoFields().Infof("Unscheduling from %s with dangling daemon set uuid %s", nodeName, dsID)
+
+						_, err = dsf.kpStore.DeletePod(kp.INTENT_TREE, nodeName, podID)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Unable to delete pod from intent tree: %v", err)
+						}
+
+						id := labels.MakePodLabelKey(nodeName, podID)
+						err = dsf.applicator.RemoveLabel(labels.POD, id, DSIDLabel)
+						if err != nil {
+							dsf.logger.NoFields().Errorf("Error removing label: %v", err)
+						}
+					}
+				}
+			}
+		}
+	}()
+}
+
+func (dsf *Farm) mainLoop(quitCh <-chan struct{}) {
+	subQuit := make(chan struct{})
+	defer close(subQuit)
+	dsWatch := dsf.dsStore.Watch(subQuit)
+
+	var changes *dsstore.WatchedDaemonSets
+	var err error
+	var ok bool
+
+	for {
+		select {
+		case <-quitCh:
+			return
+		case changes, ok = <-dsWatch:
+			if !ok {
+				return
+			}
+			if changes == nil {
+				dsf.logger.Error("Unexpected nil value received on channel")
+				return
+			}
+		}
+
+		for _, child := range dsf.children {
```

Comment:
Mind putting a comment explaining the rationale behind this loop? looks like all it does is scan all of the children for closed error channels and crashes the farm if any of them are closed. I'm curious under what circumstances you'd expect one to be closed, and why it's necessary to guarantee they're all open.